google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/SpannerPool.java

```diff
@@ -0,0 +1,407 @@
+          }
+        }
+      } else {
+        logger.log(
+            Level.WARNING,
+            "There is no Spanner registered for ConnectionOptions " + options.toString());
+      }
+    }
+  }
+
+  /**
+   * Checks that there are no {@link Connection}s that have been created by this {@link SpannerPool}
+   * that are still open, and then closes all {@link Spanner} instances in the pool. If there is at
+   * least one unclosed {@link Connection} left in the pool, the method will throw a {@link
+   * SpannerException} and no {@link Spanner} instances will be closed.
+   */
+  void checkAndCloseSpanners() {
+    checkAndCloseSpanners(CheckAndCloseSpannersMode.ERROR);
+  }
+
+  @VisibleForTesting
+  void checkAndCloseSpanners(CheckAndCloseSpannersMode mode) {
+    List<SpannerPoolKey> keysStillInUse = new ArrayList<>();
+    synchronized (this) {
+      for (Entry<SpannerPoolKey, Spanner> entry : spanners.entrySet()) {
+        if (!lastConnectionClosedAt.containsKey(entry.getKey())) {
+          keysStillInUse.add(entry.getKey());
+        }
+      }
+      if (keysStillInUse.isEmpty() || mode == CheckAndCloseSpannersMode.WARN) {
+        if (!keysStillInUse.isEmpty()) {
+          logLeakedConnections(keysStillInUse);
+          logger.log(
+              Level.WARNING,
+              "There is/are "
+                  + keysStillInUse.size()
+                  + " connection(s) still open."
+                  + " Close all connections before stopping the application");
+        }
+        closeUnusedSpanners(Long.MIN_VALUE);
+      } else {
+        logLeakedConnections(keysStillInUse);
+        throw SpannerExceptionFactory.newSpannerException(
+            ErrorCode.FAILED_PRECONDITION,
+            "There is/are "
+                + keysStillInUse.size()
+                + " connection(s) still open. Close all connections before calling closeSpanner()");
+      }
+    }
+  }
+
+  private void logLeakedConnections(List<SpannerPoolKey> keysStillInUse) {
+    synchronized (this) {
+      for (SpannerPoolKey key : keysStillInUse) {
+        for (ConnectionImpl con : connections.get(key)) {
+          if (!con.isClosed() && con.getLeakedException() != null) {
+            logger.log(Level.WARNING, "Leaked connection", con.getLeakedException());
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Closes Spanner objects that are no longer in use by connections, and where the last connection
+   * that used it was closed more than <code>closeSpannerAfterMillisecondsUnused</code> seconds ago.
+   * The delay ensures that Spanner objects are not closed unless there's a good reason for it.
+   *
+   * @param closeSpannerAfterMillisecondsUnused The number of milliseconds a {@link Spanner} object
+   *     should not have been used for a {@link Connection} before it is closed by this method.
+   */
+  @VisibleForTesting
+  void closeUnusedSpanners(long closeSpannerAfterMillisecondsUnused) {
```

Comment:
where does this parameter get its value?

Reply:
The default value that is used when nothing else is configured is found here: https://github.com/googleapis/google-cloud-java/blob/a6d9f88a154c223bb079e6e4d0d5a9619fd6b3d6/google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/SpannerPool.java#L75

This value is currently only configurable for testing. A client application cannot change this value and will always use the default.